### PR TITLE
Improve JSON support

### PR DESCRIPTION
**Pull request by njaremko**
- Improve JSON function documentation
- Add `ByteString` instances for JSON types, so query response can be used directly
- Add `Text` instances for JSON types

### CODE REVIEW

File: src/Opaleye/Internal/RunQuery.hs [hunk 1]
```diff
@@ -344,6 +386,50 @@ jsonFieldTypeParser jsonTypeName field mData = do
...
+jsonFieldTypeLazyByteParser jsonTypeName field mData = do
+    ti <- typeInfo field
+    if TI.typname ti == jsonTypeName
+       then convert
+       else returnError Incompatible field "types incompatible"
+  where
+    convert = case mData of
+        Just bs -> pure $ LBS.fromStrict bs
+        _       -> returnError UnexpectedNull field ""
+
```
Comment (tomjaguarpaw) [inline src/Opaleye/Internal/RunQuery.hs:432]: Can you remove the duplication in here by doing all the others as an `(fmap . fmap) f jsonFieldTypeByteParser`? `f` will be `IPT.strictDecodeUtf8` for `String`, for example.

**Review by tomjaguarpaw — CHANGES_REQUESTED** [review]
Great, thanks!

File: src/Opaleye/Internal/RunQuery.hs (hunk 1 above)
Comment (njaremko) [reply src/Opaleye/Internal/RunQuery.hs:432]: I can 😛 . I was being lazy.

One moment.

File: src/Opaleye/Internal/RunQuery.hs [hunk 2]
```diff
@@ -327,23 +353,51 @@ jsonFieldParser, jsonbFieldParser :: FieldParser String
...
        else returnError Incompatible field "types incompatible"
   where
     convert = case mData of
-        Just bs -> pure $ IPT.strictDecodeUtf8 bs
+        Just bs -> pure bs
         _       -> returnError UnexpectedNull field ""
 
+jsonFieldTypeLazyByteParser :: SBS.ByteString -> FieldParser LBS.ByteString
+jsonFieldTypeLazyByteParser x = (fmap . fmap . fmap) LBS.fromStrict (jsonFieldTypeByteParser x)
+
```
Comment (tomjaguarpaw) [inline src/Opaleye/Internal/RunQuery.hs:400]: Nice! Do you mind adding another `fmap` and getting rid of those `x`s? :D

Comment (njaremko) [reply src/Opaleye/Internal/RunQuery.hs]: Done 😄

**Review by tomjaguarpaw — APPROVED** [review]
Thanks! I'll get round to merging this at some point. Feel free to ping me if you want to speed me up.

**Comment by tomjaguarpaw** [comment]
Rebased by hand and merged

**Pull request closed without merging**